sensing/ahrs_ekf_mocap.cpp

```diff
@@ -0,0 +1,134 @@
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+/*******************************************************************************
+ * \file ahrs_ekf_mocap.cpp
+ *
+ * \author MAV'RIC Team
+ * \author Matthew Douglas
+ *
+ * \brief Extended Kalman Filter attitude estimation, mixing accelerometer and magnetometer
+ * x[0] : bias_x
+ * x[1] : bias_y
+ * x[2] : bias_z
+ * x[3] : q0
+ * x[4] : q1
+ * x[5] : q2
+ * x[6] : q3
+ *
+ * Takes into account the motion capture ahrs quaternion
+ *
+ ******************************************************************************/
+
+#include "sensing/ahrs_ekf_mocap.hpp"
+#include "hal/common/time_keeper.hpp"
+#include "util/kalman.hpp"
+
+extern "C"
+{
+#include "util/constants.h"
+#include "util/print_util.h"
+#include "util/maths.h"
+#include "util/vectors.h"
+#include "util/quaternions.h"
+#include "util/coord_conventions.h"
+}
+
+
+using namespace mat;
+
+//------------------------------------------------------------------------------
+// PRIVATE FUNCTIONS DECLARATION
+//------------------------------------------------------------------------------
+
+
+
+//------------------------------------------------------------------------------
+// PRIVATE FUNCTIONS IMPLEMENTATION
+//------------------------------------------------------------------------------
+
+void Ahrs_ekf_mocap::callback(Ahrs_ekf_mocap* ahrs_ekf_mocap, uint32_t sysid, mavlink_message_t* msg)
+{
+    ahrs_ekf_mocap->R_mocap_ = Mat<4, 4>(ahrs_ekf_mocap->config_.R_mocap, true);
+
+    mavlink_att_pos_mocap_t packet;
+    mavlink_msg_att_pos_mocap_decode(msg, &packet);
+
+    // Get timing
+    float t = time_keeper_get_us();
+
+    // Create matrices for the update
+    Mat<4, 1> z = Mat<4, 1>(0.0f);
+    z(0, 0) = packet.q[0];
+    z(1, 0) = packet.q[1];
+    z(2, 0) = packet.q[2];
+    z(3, 0) = packet.q[3];
+
+    Mat<4, 7> H = Mat<4, 7>(0.0f);
```

Comment:
z and H could be members of ahrs_ekf_mocap (it avoids creating them 10 times per second)